lib/elixir/lib/keyword.ex

```diff
@@ -186,6 +186,182 @@ defmodule Keyword do
+
+      iex> Keyword.validate(%{three: 3}, [one: 1, two: 2])
+      ** (ArgumentError) expected a keyword list with keys [:one, :two], got: %{three: 3}
+
+      iex> Keyword.validate([:three], [one: 1, two: 2])
+      ** (ArgumentError) expected a keyword list with keys [:one, :two], got: [:three]
+
+      iex> Keyword.validate([three: 3], [:three, 3, :two])
+      ** (ArgumentError) expected the second argument to be a list of atoms or tuples, got item: 3
+  """
+  @doc since: "1.13.0"
+  @spec validate(keyword :: keyword(), values :: [atom() | {atom(), term()}]) ::
+          {:ok, keyword}
+          | {:error, {:invalid_key, term()}}
+  def validate(keyword, values) do
+    case validate!(keyword, values, [], []) do
+      {:ok, kw} ->
+        {:ok, kw}
+
+      {:badkey, key} ->
+        {:error, {:invalid_key, key}}
+
+      :badkey ->
+        keys =
+          for value <- values do
+            if is_atom(value) do
+              value
+            else
+              elem(value, 0)
+            end
+          end
+
+        raise ArgumentError,
+              "expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"
+    end
+  end
+
+  @doc """
+  Similar to `validate/2` but returns the keyword or raises an error.
+
+  ## Examples
+      iex> Keyword.validate!([], [one: 1, two: 2]) |> Enum.sort()
+      [one: 1, two: 2]
+      iex> Keyword.validate!([two: 3], [one: 1, two: 2]) |> Enum.sort()
+      [one: 1, two: 3]
+
+  If atoms are given, they are supported as keys but do not
+  provide a default value:
+
+      iex> Keyword.validate!([], [:one, two: 2]) |> Enum.sort()
+      [two: 2]
+      iex> Keyword.validate!([one: 1], [:one, two: 2]) |> Enum.sort()
+      [one: 1, two: 2]
+
+  Passing an unknown key raises:
+
+      iex> Keyword.validate!([three: 3], [one: 1, two: 2])
+      ** (ArgumentError) unknown key :three in [three: 3], expected one of [:one, :two]
+  """
+  @doc since: "1.13.0"
+  @spec validate(keyword :: keyword(), values :: [atom() | {atom(), term()}]) ::
+          keyword | no_return
+  def validate!(keyword, values) when is_list(keyword) and is_list(values) do
+    # We use two lists to avoid reversing/concatenating
+    # lists in the middle of traversals.
+    case validate!(keyword, values, [], []) do
+      {:ok, keyword} ->
+        keyword
+
+      error ->
+        keys =
+          for value <- values do
+            if is_atom(value) do
+              value
+            else
+              elem(value, 0)
+            end
+          end
+
+        case error do
+          {:badkey, key} ->
+            raise ArgumentError,
+                  "unknown key #{inspect(key)} in #{inspect(keyword)}, " <>
+                    "expected one of #{inspect(keys)}"
+
+          :badkey ->
+            raise ArgumentError,
+                  "expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"
+        end
+    end
+  end
```

Comment:
ideally the raising variant uses the non-raising variant under the hood, I think it will remove a fair amount of duplication. We can even drop the guards as the underlying function has them already.

```elixir
  def validate!(keyword, values) do
    case validate(keyword, values) do
      {:ok, keyword} ->
        keyword

      {:error, missing_keys} ->
        # ...
    end
  end
```

regarding error, I think we have two options, perhaps something along the lines of either:

1. `"expected a keyword list with keys #{inspect(keys)}, got: #{inspect(keyword)}"`
2. `"unknown keys #{inspect(missing_keys)} in #{inspect(keyword)}, the allowed keys are: #{inspect(keys)}"`

what do you think?

Reply:
Yeah, I'll refactor the implementation then. Makes more sense, especially since both fail exactly on the same cases.

One thing I noticed is that although probably not intended `KeyError` accepts a list as `:key`. The message in itself becomes a little bit weird though, so I'll stick to ArgumentError

```elixir
iex(1)> raise KeyError, key: [:a, :b, :c]
** (KeyError) key [:a, :b, :c] not found
```